server/src/main/java/org/elasticsearch/cluster/coordination/PreVoteCollector.java

```diff
@@ -0,0 +1,192 @@
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "PreVoteCollector{" +
+            "currentLeader=" + currentLeader +
+            ", preVoteResponse=" + preVoteResponse +
+            '}';
+    }
+
+    private class PreVotingRound implements Releasable {
+        private final Set<DiscoveryNode> preVotesReceived = newConcurrentSet();
+        private final AtomicBoolean electionStarted = new AtomicBoolean();
+        private final PreVoteRequest preVoteRequest;
+        private final ClusterState clusterState;
+        private final AtomicBoolean isClosed = new AtomicBoolean();
+
+        PreVotingRound(final ClusterState clusterState, final long currentTerm) {
+            this.clusterState = clusterState;
+            preVoteRequest = new PreVoteRequest(transportService.getLocalNode(), currentTerm);
+        }
+
+        void start(final Iterable<DiscoveryNode> broadcastNodes) {
+            broadcastNodes.forEach(n -> transportService.sendRequest(n, REQUEST_PRE_VOTE_ACTION_NAME, preVoteRequest,
+                new TransportResponseHandler<PreVoteResponse>() {
+                    @Override
+                    public void handleResponse(PreVoteResponse response) {
+                        handlePreVoteResponse(response, n);
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        if (exp.getRootCause() instanceof CoordinationStateRejectedException) {
+                            logger.debug("{} failed: {}", this, exp.getRootCause().getMessage());
+                        } else {
+                            logger.debug(new ParameterizedMessage("{} failed", this), exp);
+                        }
+                    }
+
+                    @Override
+                    public String executor() {
+                        return Names.GENERIC;
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "TransportResponseHandler{" + PreVoteCollector.this + ", node=" + n + '}';
+                    }
+                }));
+        }
+
+        private void handlePreVoteResponse(final PreVoteResponse response, final DiscoveryNode sender) {
+            if (isClosed.get()) {
+                logger.debug("{} is closed, ignoring {} from {}", this, response, sender);
+                return;
+            }
+
+            // TODO the response carries the sender's current term. If an election starts then it should be in a higher term.
+
+            if (response.getLastAcceptedTerm() > clusterState.term()
+                || (response.getLastAcceptedTerm() == clusterState.term()
+                && response.getLastAcceptedVersion() > clusterState.version())) {
+                logger.debug("{} ignoring {} from {} as it is fresher", this, response, sender);
```

Comment:
why debug instead of trace logging everywhere?

Reply:
I would prefer to keep trace logging for things that are expected to happen in a stable cluster, and anything unexpected to appear at debug or higher. In a stable cluster, any election activity is unexpected.

Reply:
ok makes sense